src/project.ts

```diff
@@ -342,28 +255,64 @@ export class AtlasDataset extends BaseAtlasClass<
     return new AtlasIndex(id, this.viewer, { project: this });
   }
 
-  async delete_data(ids: string[]): Promise<void> {
-    // TODO: untested
-    // const info = await this.info
-    await this.viewer.apiCall('/v1/project/data/delete', 'POST', {
-      project_id: this.id,
-      datum_ids: ids,
-    });
+  async loadSchema() {
+    if (this.attr === undefined) {
+      await this.fetchAttributes();
+    }
+    if (this.attr === undefined) {
+      throw new Error('Attributes not found after fetching');
+    }
+    const base64Schema = this.attr.schema;
+    if (base64Schema === undefined) {
+      return null;
+    }
+    const schemaBuffer = Buffer.from(base64Schema, 'base64');
+    const array = new Uint8Array(schemaBuffer);
+    const table = tableFromIPC(array);
+    return table.schema;
   }
 
-  validate_metadata(): void {
-    // validate metadata
-  }
+  async uploadData(
+    data: Record<string, unknown> | Record<string, unknown>[]
+  ): Promise<void> {
+    const points = Array.isArray(data) ? data : [data];
+
+    const schema = await this.loadSchema();
+
+    const fields: Field[] = schema?.fields ?? [];
+
+    if (fields.length === 0) {
+      // If no schema, this is the first datapoint. We need to determine the right type
+      // for each column given our constraints
+      for (const key in points[0]) {
+        const value = points[0][key];
```

Comment:
Checking the first value won't work if that first value is null/undefined but it's actually a timestamp/number field.

Reply:
Can we bootstrap off jsarrow's inference here? 

```js
vectorFromArray(points.map(p => p[key]))
```

and then reduce the types down according to some code we already have?